### DISCUSSION TITLE
Static table name for models

### DISCUSSION THREAD

**Discussion by yanickrochon** [laravel/framework#52485] [Ideas]
While implementing a method using raw SQL queries to perform a complex database task, I was looking to get the models' table names instead of writing them in clear.

```php
$subQuery = DB::table('notifications', 'n')
   ->select(['related_element_id', 'related_element_type'])
   ->whereIn('related_element_type', [KeysTranslation::articles->value, KeysTranslation::jobs->value])
   ->toRawSql();
```

My option was to either:

1. having the table name literal: `"notifications"`
2. creating a model instance: `Notification::make()->getTable()`

Neither solution seem OK to me.

## Current implementation

After looking at the code, located in `vendor\laravel\framework\src\Illuminate\Database\Eloquent\Model.php`, I found the implementation:

```php
abstract class Model /*...*/ {

   protected $table;

   public function getTable()
   {
      return $this->table ?? Str::snake(Str::pluralStudly(class_basename($this)));
   }
}
```

and wondered why this could not be static. As in:

```php
abstract class Model /*...*/ {

   static protected $table;

   static public function getTableName()
   {
      return get_called_class()::$table ?? Str::snake(Str::pluralStudly(class_basename(get_called_class())));
   }

   public function getTable() {
      return get_called_class()::getTableName();
   }
}
```

## Example

The following example illustrates the feasability of the implementation:

```php
class Foo {
    static protected $name;

    static public function getName() {
        return get_called_class()::$name ?? class_basename(get_called_class());
    }

    public function getFoo() {
        return get_called_class()::getName();
    }
}

class Bar extends Foo {
    public function getFoo() {
        return Bar::getName();
    }
}

class Buz extends Foo {
    static protected $name = 'superbozz';
}

$foo = new Foo();
$bar = new Bar();
$buz = new Buz();

dd([$foo->getFoo(), $bar->getFoo(), $buz->getFoo()]);
// array:2 [
//   0 => "Foo"
//   1 => "Bar"
//   2 => "superbuzz
// ] 
```

## Conclusion

This idea not only preserves the existing behavior

```php
Notification::make()->getTable();
```

but enables direct static calls

```php
Notification::getTableName();
```

**Comment by henzeb** [comment]
what's wrong with the query you get from the model?

**Reply by yanickrochon** [reply]
I needed a sub-query with an alias. However, `Notification->from()` requiers a table name (!)

**Comment by ghost** [comment]
Hi. Nothing stops you from declaring it static and on construct setting it from the static. Why introduce breaking change for all?

**Reply by ghost** [reply]
Or even overwriting what you posted in an abstract model that you extend. No need to break other's implementation because you prefer it static.

**Reply by yanickrochon** [reply]
In what way is the solution I'm proposing a breaking change? Have you looked at the POC in example?

I can't seriously override every models, even Statistic's and other libraries!

The current workaround, cresting some instances, just to retrieve a table name, should never have been needed to begin with.

**Reply by ghost** [reply]
@yanickrochon I agree with "The current workaround, cresting some instances, just to retrieve a table name, should never have been needed to begin with.". But it is as it is and changing it will be a breaking change.

"I can't seriously override every models, even Statistic's and other libraries!": You can create an abstract model that extends Model and all your models extent that abstract. In that abstract you set $this->table = static::table; on construct or overwrite what you posted.
I would go for the easy solution with construct.

"In what way is the solution I'm proposing a breaking change? Have you looked at the POC in example?": Many have 
   protected $table; declared. Your solution will change that to static and I don't see it checking the protected for backward compatibility.
